[출석 체크] 출석 체크 비즈니스 로직 및 API 구성(#14)

Minjae-An opened the pull request:
## 작업 대상

<!-- 작업한 대상을 자세히 설명해주세요. -->

- 화살 송수신 내역 엔티티(`ArrowTransaction`)
- 화살 송수신 내역 Repository(`ArrowTransactionCommandRepository`)
- 출석 체크 비즈니스 로직(`ArrowService`)
- 출석 체크 API(`ArrowController`)

## 📄 작업 내용

<!-- 작업한 내용을 간단히 요약해주세요. -->

- 화살 송수신 내역에 주고 받은 화살 수(`arrows`)가 포함됨에 따라 엔티티에 필드를 추가
- 화살 송수신 내역과 관련한 생성/변경/삭제 작업을 수행하는 `Repository` 정의
- 출석 체크 비즈니스 로직 구현
    1. 이미 출석 체크된 사용자인 지 검증
    2. 화살 송수신 내역 저장
    3. 사용자 최종 접속 일시 갱신
    4. 사용자 보유 화살 개수 증가
- 출석 체크 API 구현 및 명세

## 🙋🏻 주의 사항

<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->

## 📎 관련 이슈

## 레퍼런스

<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->

Minjae-An committed:
refactor: 잘못된 클래스명 수정(#14)

Minjae-An committed:
feat: JpaAuditing 활용을 위한 설정 정의(#14)

Minjae-An committed:
feat: 화살 내역 command repository 추가(#14)

Minjae-An committed:
feat: 최근 접속 시간 갱신, 화살 증가 로직 추가(#14)

Minjae-An committed:
feat: 생성 시간, 데이터 저장시 자동 설정토록 수정(#14)

JPA Auditing 활용 생성 시간 자동 설정되도록 로직 수정

Minjae-An committed:
feat: 화살 서비스 구성 및 출석 체크 로직 추가(#14)

출석 체크 비즈니스 로직은 다음 과정을 통해 진행된다.
1. 사용자 최근 접속 시간을 통해 이미 출석 체크했는 지 확인
2. 화살 송수신 내역 저장
3. 사용자 보유 화살 개수 증가(고정 10개)

Minjae-An committed:
feat: 출석 체크 API 구현(#14)

Minjae-An committed:
feat: 화살 개수 필드 추가(#14)

화살 송수신 내역에 주고 받은 화살 개수가 포함되도록
수정

Minjae-An committed:
feat: 화살 내역 저장시 화살 수 포함하여 저장토록 수정(#14)

Minjae-An committed:
refactor: 코드 정렬 수정(#14)

Minjae-An committed:
feat: 출석 체크 API, 정상 처리시 응답 코드 202로 수정(#14)

Minjae-An committed:
Merge branch 'dev' into feat/#14-daily-check

# Conflicts:
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
refactor: 코드 정렬 수정(#14)

Minjae-An committed:
refactor: 코드 정렬 수정(#14)

Minjae-An committed:
refactor: conflict 해결을 위한 기존 코드 배치 조절(#14)

Minjae-An committed:
refactor: 코드 정렬 수정(#14)

acceptor-gyu commented on be/src/main/java/yeonba/be/arrow/controller/ArrowController.java:
```diff
@@ -32,19 +39,19 @@ public ResponseEntity<CustomResponse<UserArrowsResponse>> arrows() {
-      description = "정상적으로 출석 체크 완료"
+      responseCode = "202",
+      description = "출석 체크 정상 처리"
   )
```
```suggestion
      responseCode = "202", description = "출석 체크 정상 처리")
```

acceptor-gyu commented on be/src/main/java/yeonba/be/arrow/service/ArrowService.java:
```diff
@@ -0,0 +1,60 @@
+      throw new IllegalArgumentException("이미 출석 체크한 사용자입니다.");
+    }
+
+    ArrowTransaction arrowTransaction = new ArrowTransaction(
+        null,
+        dailyCheckUser,
+        DAILY_CHECK_ARROW_COUNT);
```
생성자를 새로 만들어서 null을 넣지 않는 것은 어떤가요?
```suggestion
    ArrowTransaction arrowTransaction = new ArrowTransaction(
        dailyCheckUser,
        DAILY_CHECK_ARROW_COUNT);
```

Minjae-An replied:
`null`을 다루는 형태보다 더 적절한 것 같습니다. 고려하여 리팩토링하겠습니다!

acceptor-gyu commented on be/src/main/java/yeonba/be/arrow/service/ArrowService.java:
```diff
@@ -0,0 +1,60 @@
+        .equals(dailyCheckDate);
+  }
+
+  private void addUserArrow(User user, int arrowCount){
+
+    user.addArrow(arrowCount);
+  }
```
이 메서드를 만든 이유는 무엇인가요?, 이 메서드를 사용하는 곳에서 `user.addArrow(arrowCount)`만 해도 될 것같은데요.

Minjae-An replied:
사용자의 화살을 증가시키는 로직이 여기 저기 퍼져있지 않고 `ArrowService`에서 일임하여 처리하는 것이 SRP를 준수하는 형태라고 판단하였습니다. 따라서 해당 로직을 메서드로 따로 추출하였으며, 향후 관련 변경에 더 유연하게 대응할 수 있는 구조라고 생각합니다.

acceptor-gyu replied:
조금 더 초기 단계부터 보면 user 클래스의 addArrow라는 함수는 단순한 dirty checking을 생각한 코드이신가요? 아니면 
user 객체의 arrow는 본인이 획득한 화살이기 때문에 스스로 본인의 화살 개수를 추가한다! 라는 관점에서 oop를 고려한 코드인가요? 아니면 또 다른 이유가 있나요?
이 부분에 대한 생각이 궁금합니다.
저였다면 후자의 이유로 작성했을 것같습니다. 오히려 arrowService에 본인의 역할을 넘기면서 본인의 주체성을 떨어뜨리는 것처럼 느껴져서요.
그리고 현재 user의 화살 개수를 증가시킬 때 코드가 분산되어 있다는 말이 잘 이해가 되지 않습니다. arrowService.addUserArrow()와 user.addArrow()는 제가 봤을때는 똑같아 보여서요.

Minjae-An replied:
`addArrow` 라는 메서드는 oop를 고려하여 작성한 코드입니다.다만 특정 사용자에게 화살 증가를 요청하는 작업을
책임 측면에서 `ArrowService`가 일임하여 처리하는 것이 맞다고 생각하였습니다. '본인의 주체성을 떨어뜨린다'는 
측면에서 해당 의견이 더 설득력 있다고 판단됩니다. 반영하여 리팩토링 하겠습니다!

acceptor-gyu reviewed (approved):
확인 후, 논의할 부분 있으면 논의하고 머지해주세요.

Minjae-An committed:
feat: 보낸 사용자를 포함하지 않는 생성자 추가(#14)

출석 체크시 보낸 사용자가 null로 설정해야 되는 상황에서
사용할 수 있는 별도의 생성자 추가

Minjae-An committed:
refactor: 출석 체크 비즈니스 로직 수정(#14)

- id 기반으로 사용자를 조회하도록 수정
- 보낸 사용자(receivedUser)를 설정하지 않는 ArrowTransaction 생성자 사용

Minjae-An committed:
refactor: 출석체크 API 로직 수정(#14)

- 요청 속성으로 사용자 ID를 받도록 로직 수정
- API 명세 속성 한 라인에 오도록 코드 배치 수정

Minjae-An committed:
refactor: 사용자 화살 증가, User가 처리하도록 로직 수정(#14)

클래스 책임 측면에서 사용자 화살 증가를 User가 일임하여 처리하도록
로직 수정
